importer/xmlparser/src/test/java/org/itsallcode/openfasttrace/importer/xmlparser/tree/TreeElementTest.java

```diff
@@ -0,0 +1,121 @@
+import static org.mockito.ArgumentMatchers.same;
+import static org.mockito.Mockito.*;
+
+import java.util.function.Consumer;
+
+import org.itsallcode.openfasttrace.api.core.Location;
+import org.itsallcode.openfasttrace.importer.xmlparser.event.StartElementEvent;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InOrder;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.xml.sax.Attributes;
+
+@ExtendWith(MockitoExtension.class)
+class TreeElementTest {
+    private static final String ELEMENT_URI = "uri";
+    private static final String ELEMENT_LOCAL_NAME = "localName";
+    private static final String ELEMENT_QNAME = "qname";
+
+    @Mock
+    Consumer<TreeElement> listenerMock1;
+    @Mock
+    Consumer<TreeElement> listenerMock2;
+
+    @Test
+    void testToString() {
+        assertThat(testee().toString(), equalTo(
+                "TreeElement [element=StartElementEvent [qName={uri}localName, attributeMap={qname0=Attribute [qName=qname0, value=value0]}, location=path:2], characterData=, endElementListeners=[], parent=null]"));
+    }
+
+    @Test
+    void getElement() {
+        assertThat(testee().getElement().getName().getLocalPart(), equalTo(ELEMENT_LOCAL_NAME));
+    }
+
+    @Test
+    void getCharacterDataEmpty() {
+        assertThat(testee().getCharacterData(), equalTo(""));
+    }
+
+    @Test
+    void getCharacterData() {
+        final TreeElement element = testee();
+        element.addCharacterData("chars");
+        assertThat(element.getCharacterData(), equalTo("chars"));
+    }
+
+    @Test
+    void getCharacterDataMultipleAdded() {
+        final TreeElement element = testee();
+        element.addCharacterData("chars");
+        element.addCharacterData(", more chars");
+        assertThat(element.getCharacterData(), equalTo("chars, more chars"));
+    }
+
+    @Test
+    void isRootTrueForNullParent() {
+        assertThat(testee().isRootElement(), is(true));
+    }
+
+    @Test
+    void isRootFalsForNonNullParent() {
+        final TreeElement element = new TreeElement(createStartElement(), mock(TreeElement.class));
+        assertThat(element.isRootElement(), is(false));
+    }
+
+    @Test
+    void getAttributeValueByName() {
+        assertThat(testee().getAttributeValueByName("qname0").getValue(), equalTo("value0"));
+    }
+
+    @Test
+    void getLocation() {
+        assertThat(testee().getLocation().getPath(), equalTo("path"));
+    }
+
+    @Test
+    void invokeEndElementListenersSuccedsForEmptyList() {
+        assertDoesNotThrow(testee()::invokeEndElementListeners);
+    }
+
+    @Test
+    void invokeEndElementListenersCallsSingleEntry() {
+        final TreeElement element = testee();
+        element.addEndElementListener(listenerMock1);
+        element.invokeEndElementListeners();
+        verify(listenerMock1).accept(same(element));
```

Comment:
```suggestion
        verify(listenerMock1).accept(same(element));
```
Isn't the fact that it is the same instance testing for an implementation detail? Hm. Element is probably supposed to be mutable in most cases, so it might be okay as a test.

Reply:
Correct, this is testing an implementation detail. I am willing to accept the risk of having to change this test should the implementation change.
The `element` object is immutable.